qiskit/algorithms/eigen_solvers/eigen_solver.py

```diff
@@ -13,12 +13,16 @@
 """The Eigensolver interface"""
 
 from abc import ABC, abstractmethod
-from typing import List, Optional
+from typing import Dict, Optional, Any, List, Union, TypeVar
 
 import numpy as np
 from qiskit.opflow import OperatorBase
 from ..algorithm_result import AlgorithmResult
 
+# Introduced new type to maintain readability.
+T = TypeVar('T')
+ListOrDict = Union[List[Optional[T]], Dict[Any, T]]
```

Comment:
I think the following should be fine in this scenario:
```suggestion
ListOrDict = Union[List[Optional[T]], Dict[str, T]]
```

Reply:
The author originally had `str` as a typehint for key for dict. I saw no real reason why we needed to limit the user to requiring to provide strings as keys. After all I may find different keys more convenient for my application/usage so in my mind there was no need to limit this.